### PR TITLE
feat: added HTTP capability

### LINKED ISSUE #96: create a http server capability

### PR DESCRIPTION
**Pull request by Mossaka**
This PR adds HTTP server capability to `slight`, it's main contributions are
1. `wit/http-api.wit` HTTP API WIT file. 
1. `/crates/http-api` adds a host implementation for HTTP server
2. `/crates/http-handler-macro` adds macro for registering HTTP handler functions
3. `examples/http-demo` adds an guest example for using HTTP WIT

It will close #96

### CODE REVIEW

File: crates/http/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+[package]
+name = "http"
+version = "0.1.0"
+edition = "2021"
+authors = ["DeisLabs Engineering Team"]
+
+
+[lib]
+doctest = false
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply crates/http/Cargo.toml:9]: We have some unittests in `http` crate. I don't think we should toggle off `test`.

File: crates/http/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+#[derive(Clone, Debug)]
+struct Route {
+    method: Methods,
+    route: String,
+    handler: String,
+}
+
+/// A Router implementation for the HTTP interface
+#[derive(Default, Clone, Debug)]
+pub struct RouterProxy {
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply crates/http/src/lib.rs:48]: I guess it's fine for now. both `http` and `events` do not have multiple components.

File: crates/http/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    /// The root directory of the filesystem
+    _base_uri: String,
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply crates/http/src/lib.rs:50]: Not used anywhere.

File: .gitignore [hunk 4]
```diff
@@ -1,3 +1,6 @@
 target/
 .vscode/
-bindings.rs
\ No newline at end of file
+bindings.rs
+.idea
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply .gitignore:4]: generated files for other IDEA IDE

File: crates/http/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (danbugs) [reply crates/http/src/lib.rs:48]: I meant mostly for convention of calling these structures `SomethingInner`~ up to u, tho

**Review by danbugs — CHANGES_REQUESTED** [review]
mostly, looking good to me~ just bringing back a couple of resolved items and small nits

**Comment by Mossaka** [comment]
I just pushed a commit that adds `put` `post` and `delete` methods to our http-handler.

File: crates/http/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mossaka) [reply crates/http/src/lib.rs:51]: Renamed~ 
\

File: crates/http/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        tokio::task::spawn(graceful);
+
+        let arc_tx = Arc::new(Mutex::new(tx));
+        self.host_state.closer = Some(arc_tx.clone());
+
+        Ok(ServerProxy { closer: arc_tx })
+    }
+
+    fn server_stop(&mut self, server: &Self::Server) -> Result<(), Error> {
+        let clone = server.clone();
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply crates/http/src/lib.rs:297]: done~

**Review by danbugs — APPROVED** [review]
LGTM~ I think you just forgot to address one of the comment typo nits, I brought up!

File: crates/http-api/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+use hyper::{
+    body::HttpBody as HyperHttpBody,
+    header::{HeaderName, HeaderValue},
+    Body, HeaderMap, StatusCode,
+};
+
+wit_bindgen_wasmtime::import!("../../wit/http-handler.wit");
+wit_error_rs::impl_error!(Error);
+
+/// A owned http_handler::HeadersParam
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply crates/http-api/src/lib.rs:12]: I will resolve this in `http-builder` branch

**Pull request merged**